server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -14,30 +14,41 @@
 import org.elasticsearch.search.fetch.FetchContext;
 import org.elasticsearch.search.fetch.FetchSubPhase;
 import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
 import org.elasticsearch.search.fetch.StoredFieldsSpec;
 
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.Map;
 
 /**
  * A fetch sub-phase for high-level field retrieval. Given a list of fields, it
- * retrieves the field values through the relevant {@link org.elasticsearch.index.mapper.ValueFetcher}
- * and returns them as document fields.
+ * retrieves the field values and returns them as document fields.
  */
 public final class FetchFieldsPhase implements FetchSubPhase {
+
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
-        FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
-        if (fetchFieldsContext == null) {
-            return null;
-        }
-
-        FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fetchFieldsContext.fields());
+        final FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
+        final StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+        final List<FieldAndFormat> fieldAndFormatList = fetchFieldsContext == null ? Collections.emptyList() : fetchFieldsContext.fields();
+        final FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fieldAndFormatList);
```

Comment:
I think that you don't want to call FieldFetcher if you have no fields to fetch? The conditional that you have removed above. I do see how you want to do the rest here as this phase needs now to also fetch metadata fields.

Reply:
Looking deeper, I'd suggest making this change separately: moving the fetch of metadata fields out from StoredFieldsPhase

Reply:
Basically I moved fetching metadata fields here no matter if they are stored or use doc values. I understand that this can be moved to another phase, let's say `FecthMetadataFields`. ANyway from a functional point of view I would like to focus first on getting bwc work. Ideally once those are ok extracting this into another phase should not be an issue...